### PR TITLE
Fix data race in LocalServer

### LINKED ISSUE #62025: Data race on `Context::setSettings` & `Context::getReadSettings`
https://s3.amazonaws.com/clickhouse-test-reports/61163/af66e1f7c49e86017b0d71488b4509d450522a71/stateless_tests__tsan__[5_5]/run.log

```
WARNING: ThreadSanitizer: data race (pid=203585)
  Read of size 1 at 0x7b9c00008bbf by thread T26:
    #0 std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::__is_long[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/string:1499:33 (clickhouse+0x1636977f) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #1 std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::__get_pointer[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/string:1620:17 (clickhouse+0x1636977f)
    #2 std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::data[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/string:1279:73 (clickhouse+0x1636977f)
    #3 std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::operator std::__1::basic_string_view<char, std::__1::char_traits<char>>[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/string:903:65 (clickhouse+0x1636977f)
    #4 DB::Context::getReadSettings() const build_docker/./src/Interpreters/Context.cpp:5173:40 (clickhouse+0x1636977f)
    #5 DB::MergeTreeSequentialSource::MergeTreeSequentialSource(DB::MergeTreeSequentialSourceType, DB::MergeTreeData const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>, std::__1::optional<DB::MarkRanges>, bool, bool, bool, bool) build_docker/./src/Storages/MergeTree/MergeTreeSequentialSource.cpp:151:43 (clickhouse+0x18fecb71) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #6 DB::MergeTreeSequentialSource* std::__1::construct_at[abi:v15000]<DB::MergeTreeSequentialSource, DB::MergeTreeSequentialSourceType&, DB::MergeTreeData const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, std::__1::shared_ptr<DB::IMergeTreeDataPart const>&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>&, std::__1::optional<DB::MarkRanges>, bool, bool&, bool&, bool&, DB::MergeTreeSequentialSource*>(DB::MergeTreeSequentialSource*, DB::MergeTreeSequentialSourceType&, DB::MergeTreeData const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, std::__1::shared_ptr<DB::IMergeTreeDataPart const>&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>&, std::__1::optional<DB::MarkRanges>&&, bool&&, bool&, bool&, bool&) build_docker/./contrib/llvm-project/libcxx/include/__memory/construct_at.h:35:48 (clickhouse+0x18ff05b3) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #7 void std::__1::allocator_traits<std::__1::allocator<DB::MergeTreeSequentialSource>>::construct[abi:v15000]<DB::MergeTreeSequentialSource, DB::MergeTreeSequentialSourceType&, DB::MergeTreeData const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, std::__1::shared_ptr<DB::IMergeTreeDataPart const>&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>&, std::__1::optional<DB::MarkRanges>, bool, bool&, bool&, bool&, void, void>(std::__1::allocator<DB::MergeTreeSequentialSource>&, DB::MergeTreeSequentialSource*, DB::MergeTreeSequentialSourceType&, DB::MergeTreeData const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, std::__1::shared_ptr<DB::IMergeTreeDataPart const>&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>&, std::__1::optional<DB::MarkRanges>&&, bool&&, bool&, bool&, bool&) build_docker/./contrib/llvm-project/libcxx/include/__memory/allocator_traits.h:298:9 (clickhouse+0x18fef422) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #8 std::__1::__shared_ptr_emplace<DB::MergeTreeSequentialSource, std::__1::allocator<DB::MergeTreeSequentialSource>>::__shared_ptr_emplace[abi:v15000]<DB::MergeTreeSequentialSourceType&, DB::MergeTreeData const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, std::__1::shared_ptr<DB::IMergeTreeDataPart const>&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>&, std::__1::optional<DB::MarkRanges>, bool, bool&, bool&, bool&>(std::__1::allocator<DB::MergeTreeSequentialSource>, DB::MergeTreeSequentialSourceType&, DB::MergeTreeData const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, std::__1::shared_ptr<DB::IMergeTreeDataPart const>&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>&, std::__1::optional<DB::MarkRanges>&&, bool&&, bool&, bool&, bool&) build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:292:9 (clickhouse+0x18fef422)
    #9 std::__1::shared_ptr<DB::MergeTreeSequentialSource> std::__1::allocate_shared[abi:v15000]<DB::MergeTreeSequentialSource, std::__1::allocator<DB::MergeTreeSequentialSource>, DB::MergeTreeSequentialSourceType&, DB::MergeTreeData const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, std::__1::shared_ptr<DB::IMergeTreeDataPart const>&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>&, std::__1::optional<DB::MarkRanges>, bool, bool&, bool&, bool&, void>(std::__1::allocator<DB::MergeTreeSequentialSource> const&, DB::MergeTreeSequentialSourceType&, DB::MergeTreeData const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, std::__1::shared_ptr<DB::IMergeTreeDataPart const>&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>&, std::__1::optional<DB::MarkRanges>&&, bool&&, bool&, bool&, bool&) build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:953:55 (clickhouse+0x18fef422)
    #10 std::__1::shared_ptr<DB::MergeTreeSequentialSource> std::__1::make_shared[abi:v15000]<DB::MergeTreeSequentialSource, DB::MergeTreeSequentialSourceType&, DB::MergeTreeData const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, std::__1::shared_ptr<DB::IMergeTreeDataPart const>&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>&, std::__1::optional<DB::MarkRanges>, bool, bool&, bool&, bool&, void>(DB::MergeTreeSequentialSourceType&, DB::MergeTreeData const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, std::__1::shared_ptr<DB::IMergeTreeDataPart const>&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>&, std::__1::optional<DB::MarkRanges>&&, bool&&, bool&, bool&, bool&) build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:962:12 (clickhouse+0x18fef422)
    #11 DB::createMergeTreeSequentialSource(DB::MergeTreeSequentialSourceType, DB::MergeTreeData const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>, std::__1::optional<DB::MarkRanges>, bool, bool, bool, bool, std::__1::shared_ptr<std::__1::atomic<unsigned long>>) build_docker/./src/Storages/MergeTree/MergeTreeSequentialSource.cpp:317:31 (clickhouse+0x18fef422)
    #12 DB::MergeTask::ExecuteAndFinalizeHorizontalPart::createMergedStream() build_docker/./src/Storages/MergeTree/MergeTask.cpp:939:21 (clickhouse+0x18d6bd9c) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #13 DB::MergeTask::ExecuteAndFinalizeHorizontalPart::prepare() build_docker/./src/Storages/MergeTree/MergeTask.cpp:334:5 (clickhouse+0x18d68352) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #14 DB::MergeTask::ExecuteAndFinalizeHorizontalPart::execute() build_docker/./src/Storages/MergeTree/MergeTask.cpp:438:9 (clickhouse+0x18d6f97b) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #15 DB::MergeTask::execute() build_docker/./src/Storages/MergeTree/MergeTask.cpp:889:29 (clickhouse+0x18d767a8) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #16 DB::MergePlainMergeTreeTask::executeStep() build_docker/./src/Storages/MergeTree/MergePlainMergeTreeTask.cpp:57:33 (clickhouse+0x1927676d) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #17 DB::MergeTreeBackgroundExecutor<DB::DynamicRuntimeQueue>::routine(std::__1::shared_ptr<DB::TaskRuntimeData>) build_docker/./src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:282:42 (clickhouse+0x18d92148) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #18 DB::MergeTreeBackgroundExecutor<DB::DynamicRuntimeQueue>::threadFunction() build_docker/./src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:348:13 (clickhouse+0x18d93054) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #19 DB::MergeTreeBackgroundExecutor<DB::DynamicRuntimeQueue>::MergeTreeBackgroundExecutor(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, unsigned long, unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, std::__1::basic_string_view<char, std::__1::char_traits<char>>)::'lambda'()::operator()() const build_docker/./src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:57:47 (clickhouse+0x18d952e1) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #20 decltype(std::declval<DB::MergeTreeBackgroundExecutor<DB::DynamicRuntimeQueue>::MergeTreeBackgroundExecutor(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, unsigned long, unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, std::__1::basic_string_view<char, std::__1::char_traits<char>>)::'lambda'()&>()()) std::__1::__invoke[abi:v15000]<DB::MergeTreeBackgroundExecutor<DB::DynamicRuntimeQueue>::MergeTreeBackgroundExecutor(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, unsigned long, unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, std::__1::basic_string_view<char, std::__1::char_traits<char>>)::'lambda'()&>(DB::MergeTreeBackgroundExecutor<DB::DynamicRuntimeQueue>::MergeTreeBackgroundExecutor(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, unsigned long, unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, std::__1::basic_string_view<char, std::__1::char_traits<char>>)::'lambda'()&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x18d952e1)
    #21 void std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::MergeTreeBackgroundExecutor<DB::DynamicRuntimeQueue>::MergeTreeBackgroundExecutor(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, unsigned long, unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, std::__1::basic_string_view<char, std::__1::char_traits<char>>)::'lambda'()&>(DB::MergeTreeBackgroundExecutor<DB::DynamicRuntimeQueue>::MergeTreeBackgroundExecutor(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, unsigned long, unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, std::__1::basic_string_view<char, std::__1::char_traits<char>>)::'lambda'()&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9 (clickhouse+0x18d952e1)
    #22 std::__1::__function::__default_alloc_func<DB::MergeTreeBackgroundExecutor<DB::DynamicRuntimeQueue>::MergeTreeBackgroundExecutor(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, unsigned long, unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, std::__1::basic_string_view<char, std::__1::char_traits<char>>)::'lambda'(), void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0x18d952e1)
    #23 void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::MergeTreeBackgroundExecutor<DB::DynamicRuntimeQueue>::MergeTreeBackgroundExecutor(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, unsigned long, unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, std::__1::basic_string_view<char, std::__1::char_traits<char>>)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16 (clickhouse+0x18d952e1)
    #24 std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16 (clickhouse+0xebc9d0e) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #25 std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12 (clickhouse+0xebc9d0e)
    #26 ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__1::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) build_docker/./src/Common/ThreadPool.cpp:455:13 (clickhouse+0xebc9d0e)
    #27 void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:217:73 (clickhouse+0xebce0dc) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #28 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0xebce0dc)
    #29 decltype(auto) std::__1::__apply_tuple_impl[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&, std::__1::tuple<>&>(void&&, std::__1::tuple<>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/tuple:1789:1 (clickhouse+0xebce0dc)
    #30 decltype(auto) std::__1::apply[abi:v15000]<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()&, std::__1::tuple<>&>(void&&, std::__1::tuple<>&) build_docker/./contrib/llvm-project/libcxx/include/tuple:1798:1 (clickhouse+0xebce0dc)
    #31 ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()::operator()() build_docker/./src/Common/ThreadPool.h:223:13 (clickhouse+0xebce0dc)
    #32 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0xebce041) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #33 void std::__1::__invoke_void_return_wrapper<void, true>::__call<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&>(ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'()&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9 (clickhouse+0xebce041)
    #34 std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0xebce041)
    #35 void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16 (clickhouse+0xebce041)
    #36 std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16 (clickhouse+0xebc5fee) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #37 std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12 (clickhouse+0xebc5fee)
    #38 ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) build_docker/./src/Common/ThreadPool.cpp:455:13 (clickhouse+0xebc5fee)
    #39 void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:217:73 (clickhouse+0xebcbeb1) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #40 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0xebcbeb1)
    #41 void std::__1::__thread_execute[abi:v15000]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(std::__1::tuple<void, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/thread:284:5 (clickhouse+0xebcbeb1)
    #42 void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*) build_docker/./contrib/llvm-project/libcxx/include/thread:295:5 (clickhouse+0xebcbeb1)

  Previous write of size 8 at 0x7b9c00008bb8 by main thread:
    #0 __tsan_memcpy <null> (clickhouse+0x6fffb77) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #1 std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::operator=(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) FunctionFQDN.cpp (clickhouse+0x709bba5) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #2 DB::SettingFieldString::operator=(DB::SettingFieldString const&) build_docker/./src/Core/SettingsFields.h:220:8 (clickhouse+0x162f8b26) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #3 DB::SettingsTraits::Data::operator=(DB::SettingsTraits::Data const&) build_docker/./src/Core/Settings.h:1216:1 (clickhouse+0x162f8b26)
    #4 DB::BaseSettings<DB::SettingsTraits>::operator=(DB::BaseSettings<DB::SettingsTraits> const&) build_docker/./src/Core/BaseSettings.h:56:60 (clickhouse+0x163462dc) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #5 DB::Settings::operator=(DB::Settings const&) build_docker/./src/Core/Settings.h:1222:8 (clickhouse+0x163462dc)
    #6 DB::Context::setSettings(DB::Settings const&) build_docker/./src/Interpreters/Context.cpp:2068:14 (clickhouse+0x163462dc)
    #7 DB::ClientBase::adjustSettings() build_docker/./src/Client/ClientBase.cpp:722:21 (clickhouse+0x1939bf2f) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #8 DB::LocalServer::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) build_docker/./programs/local/LocalServer.cpp:472:5 (clickhouse+0xeda9b6d) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #9 Poco::Util::Application::run() build_docker/./base/poco/Util/src/Application.cpp:315:8 (clickhouse+0x1cfc681e) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #10 mainEntryClickHouseLocal(int, char**) build_docker/./programs/local/LocalServer.cpp:895:20 (clickhouse+0xedb6d0c) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)
    #11 main build_docker/./programs/main.cpp:498:21 (clickhouse+0x708c3d8) (BuildId: 88ed8c597ae708bc25d62d74e82a1842e941d502)

```

### PR DESCRIPTION
**Pull request by evillique**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Closes https://github.com/ClickHouse/ClickHouse/issues/62025.

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 6aad90db0ff3a8a2c83adb1597905105e79a076f with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62309/6aad90db0ff3a8a2c83adb1597905105e79a076f/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Interpreters/Context.cpp [hunk 1]
```diff
@@ -5170,69 +5170,71 @@ ReadSettings Context::getReadSettings() const
 {
     ReadSettings res;
 
-    std::string_view read_method_str = settings.local_filesystem_read_method.value;
+    auto settings_copy = getSettings();
```
Comment (antonio2368) [inline src/Interpreters/Context.cpp:5173]: why not just
```cpp
SharedLockGuard lock(&mutex_: mutex);
```

Comment (Avogar) [reply src/Interpreters/Context.cpp:5173]: I actually have another question, data race is between thread in clickhouse-local that modifies global context settings after it’s initialisation and some background thread in MergeTree. This sounds not right, background threads should be started after all modifications an initialisations in global context are done. So maybe there is a bug in clickhouse-local with initialisation order

**Comment by Avogar** [comment]
The actual problem is that `adjustSettings` in `LocalServer` happens when we already started async tables loading that can use global context. I think we should call `adjustSettings` in `applySettingsOverridesForLocal` because it happens before tables loading is started

**Comment by Avogar** [comment]
@evillique will you make a proper fix I described? If you don't have time, I can do it

**Comment by evillique** [comment]
@Avogar Sorry, I got distracted and forgot about this PR. I will make a proper fix

**Review by Avogar — APPROVED** [review]

**Pull request merged**